windows/src/engine/kmtip/kmkey.cpp

```diff
@@ -243,10 +264,51 @@ HRESULT WINAPI CKeymanEditSession::KeymanGetContext(int n, PWSTR buf)   // I3567
     return S_FALSE;
   }
 
-  if (!GetLeftOfSelection(_ec, _pContext, buf, n)) {   // I4933
+  if (!GetLeftOfSelection(_ec, _pContext, buf, n, &hrGetSelection, &cFetched, &tfSelection)) {  // I4933
     return S_FALSE;   // I4933
   }
 
+  // now check for selected text
+  if (!SUCCEEDED(hr = KeymanIsTextSelected(hrGetSelection, cFetched, tfSelection, isTextSelected))) {
+    Log(L"KeymanGetContext: Warning: KeymanIsTextSelected failed");
+    // Continue to return S_OK, even though checking if a text selection exists has failed.
+    // Reaching this point means the context in 'buf' is valid and will be passed to the caller.
+    // isTextSelected will be false, which is better than returning E_FAIL at this point.
+  }
+  return S_OK;
+}
+
+HRESULT
+CKeymanEditSession::KeymanIsTextSelected(HRESULT hrGetSelection, ULONG cFetched, TF_SELECTION tfSelection, BOOL *isTextSelected) {
+  LogEnter();
+
+  HRESULT hr;
+
+  *isTextSelected = FALSE;
+
+  if (!SUCCEEDED(hrGetSelection)) {
+    return E_FAIL;
+  }
+  if (hrGetSelection == TF_E_NOSELECTION) {
+    // No selection is present
+    return S_OK;
+  }
+
+  if (cFetched > 0) {
+    // check if the range is empty
+    BOOL isEmpty = FALSE;
+
+    // Compare the start and end of the range
+    if (!SUCCEEDED(hr = tfSelection.range->IsEmpty(_ec, &isEmpty))) {
+      Log(L"KeymanIsTextSelected: Exit: Testing range->isEmpty");
+      return hr;
+    }
+
+    if (!isEmpty) {
+      *isTextSelected = TRUE;
+    }
+    tfSelection.range->Release();
```

Comment:
This shouldn't be here -- it doesn't correspond to acquisition of the range. It should be done in the caller -- even though it is implicitly calling `tfSelection.range->AddRef()` due to the call to `GetLeftOfSelection`, as least that clarifies the lifecycle due to the variable being local to that function. Also helps where range is not released on line 303 above.